### ISSUE TITLE
Data sampling seems to be more complicated than necessary

### ISSUE THREAD

**Issue by IssamLaradji** [pytorch/pytorch#3356]
I love Pytorch for its flexibility and debugging friendly environment, and like Andrej Karpathy said after using Pytorch, "I have more energy. My skin is clearer. My eye sight has improved."

However, I am finding sampling from datasets a bit more convoluted that it might need to be. I was hoping for a way to efficiently extract samples (using multiprocessing) from the dataset by providing a batch index list as input, e.g. `batch_indices = [5,4,2]; loader.get_indices(batch_indices)`, where `loader`is a `DataLoader` object. 

Currently, if I want to sample using a non-uniform distribution, first I have to define a sampler class for the `loader`, then within the class I have to define a generator that returns indices from a pre-defined list. Later, whenever the sampling distribution changes I have to re-create the sampler object that takes input values which are used to compute the new sampling distribution.

As a result, defining the data loader would be something like,
```python
loader = data.DataLoader(train_set, sampler=sampler(train_set), num_workers=2)`
```
for the initialization step; and
```python
loader.sampler = sampler(train_set, some_values)
```
between epochs.

This seems to give me certain restrictions (please correct me if I am wrong),
1. dynamic sampling is not well supported with this approach - consider the case where the sampling distribution or the batch size changes after every iteration (not epoch); and

2. this makes it necessary to have an epoch-based outer loop. Once I tried to avoid having epochs by using `itertools.cycle` on a `DataLoader` with `RandomSampler`, but it gave me a bad memory leak.

Therefore, is it an issue if the API allows for a sampling procedure that looks like the code below ? My goal is to have a flexible data sampling interface while harnessing the multiprocessing power in `DataLoader`.

```python
for i in range(n_iters):
    # 1. Sample according to `probs`
    indices = np.random.choice(n, batch_size, p=probs)
    batch = loader.get_indices(indices)
   
    # 2. Optimization step
    opt.zero_grad()
    loss = model.compute_loss(batch)
    loss.backward()
    opt.step()

    # 3. Change probs values according to some criterion
    probs = get_newProbs(probs, loss)
```

**Comment by apaszke** [comment]
Ad 1. Dynamic sampling is well supported and there's no reason to have to re-assign the `.sampler` attribute. Creating `DataLoader`s is very cheap (only instantiating their iterators is more expensive). Your sampler can return itself as an iterator, and you can make it a mutable object. Consider this example:

```python
my_sampler = DynamicSampler()
loader = DataLoader(dataset, sampler=my_sampler)
for batch in loader:
    result = ... # do sth with the batch
    my_sampler.update_distribution(result)
    if result.should_stop:
        my_sampler.finish_sequence() # calling `next(my_sampler)` will raise `StopIteration` and terminate the loop
```

Regarding 2., it would be really helpful for us if you could provide us a small repro for the memory leak.

I don't think that adding worker threads to datasets is a good idea. It will make it considerably harder for people to write them, and it makes the semantics of sharing datasets with worker processes dangerous (forking with threads is a bad idea in general). Datasets are just lazy lists, and they should remain like this. If you need multithreaded data loading, this should be a different abstraction, but as far as I see everything you mentioned seems possible to implement with `DataLoader`.

**Comment by IssamLaradji** [comment]
Thanks a lot for explaining! This sheds light on the causes of the issues I was facing with sampling.
It seems I am getting much closer to resolving one of my main use cases. 

Consider the use case where we have to carefully select a batch at each iteration. So I compute a score for each sample and find that the best next batch to pick has indices [5,6,7].

Now, I would like to select this in my next iteration, would something like this be efficient ?
The only things I see that might affect performance are the two calls I make with `iter` but it might be negligible.

```python
# 1. DATASET
class DummyDataset(Dataset):
    def __init__(self):
        pass

    def __getitem__(self, index):
        return torch.ones(5)*index

    def __len__(self):
        return 100

# 2. SAMPLER
class DynamicSampler(object):
    def __init__(self):
        self.next_batch = [0]

    def next_sample(self, indList):
        self.next_batch = indList

    def __iter__(self):
        return iter(self.next_batch)

    def __len__(self):
        return 100

# EXAMPLE
if __name__ == '__main__':
    dataset = DummyDataset()
    sampler = DynamicSampler()
    loader = DataLoader(dataset, sampler=sampler)

    best_batch = [5, 6, 7, 10, 11]
    for i in range(n_iters):
        # Set the next batch indices as the best batch
        loader.next_sample(best_batch)

        # Get Batch
        loader.batch_sampler.batch_size = len(best_batch)
        batch = iter(loader).next()
        
        # Update weights
        model.partial_fit(batch)
       
        # Get the next best batch
        best_batch = get_nextBestBatch(model, dataset)

```

PS: I will try to reproduce the memory leak problem I had and post it as a new issue.

**Comment by apaszke** [comment]
I see. This might be slow in general, so we didn't anticipate anyone doing this. Right now you have a sequential dependency, which is that you can only start loading the next batch once you already finish processing the previous one, and you can't do this multiple steps ahead, which means that multiprocessing won't help you a lot. I'd anticipate that data loading would be a major bottleneck for your training in this regime. A simple workaround would be to index the dataset directly, and make `__getitem__` accept lists of indices as `index`

**Comment by ozancaglayan** [comment]
I think if the samplers were producing batch indices at once instead of sequentially, it would be much easier to implement some advanced iteration mechanisms. I would like to implement sorted batches w.r.t target sequence lengths for example and right now none of the provided building blocks suit me.

The fact that the samplers are returning each index sequentially is also time-consuming when you slice a big numpy tensor and then merge it using the default or custom collate_fn although a numpy tensor is sliceable with a set of indices directly.

**Comment by jekbradbury** [comment]
> sorted batches w.r.t target sequence lengths
This is the default for https://github.com/pytorch/text, but I'm guessing you have a use case that doesn't work well with torchtext for other reasons. Let us know if there's something we can fix on that side though.

**Comment by ozancaglayan** [comment]
Main motivation was to learn PyTorch by coding every necessary stuff. I first checked `torchtext` but did not want to bring it in as a dependency which brings along nltk and spacy. It also seemed to have many new layers of abstractions like Samples, Fields. Moreover, with images and texts I thought that at some point I would certainly need to code something ad-hoc which may not be covered by any provided facility. But thanks, I'm still checking it from time to time and maybe at some point I'll consider switching to it :)

**Comment by apaszke** [comment]
@ozancaglayan Note that what you wanted is currently possible: `DataLoader`s accept a `batch_sampler` argument: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L258-L260

**Comment by ozancaglayan** [comment]
Yes I've noticed this. I think the best way is to write a custom `BatchSampler`. Maybe the original question in this thread also can be implemented with a custom BatchSampler and some `yield` logic to pass the previous batch scores so that the sampler can directly compute the new indices.

**Comment by jekbradbury** [comment]
(also, it's far from obvious since they're in the requirements.txt, but NLTK and SpaCy are optional dependencies and aren't installed if you just use `pip install torchtext`)

**Comment by ozancaglayan** [comment]
@IssamLaradji I've coded a `BatchSampler` which from time to time updates its own scores for each sample and resamples according to that distribution at each iteration. Maybe it can be useful for you.

https://gist.github.com/ozancaglayan/8ff67259f73e18be078177f1ae5a7805

**Comment by IssamLaradji** [comment]
Thanks @ozancaglayan  - I will give it a try today!

**Comment by kris-singh** [comment]
@IssamLaradji Look at my implementation [here] (https://github.com/pytorch/pytorch/issues/7359)

**Comment by Tsingularity** [comment]
> Ad 1. Dynamic sampling is well supported and there's no reason to have to re-assign the `.sampler` attribute. Creating `DataLoader`s is very cheap (only instantiating their iterators is more expensive). Your sampler can return itself as an iterator, and you can make it a mutable object. Consider this example:
> 
> ```python
> my_sampler = DynamicSampler()
> loader = DataLoader(dataset, sampler=my_sampler)
> for batch in loader:
>     result = ... # do sth with the batch
>     my_sampler.update_distribution(result)
>     if result.should_stop:
>         my_sampler.finish_sequence() # calling `next(my_sampler)` will raise `StopIteration` and terminate the loop
> ```
> 
> Regarding 2., it would be really helpful for us if you could provide us a small repro for the memory leak.
> 
> I don't think that adding worker threads to datasets is a good idea. It will make it considerably harder for people to write them, and it makes the semantics of sharing datasets with worker processes dangerous (forking with threads is a bad idea in general). Datasets are just lazy lists, and they should remain like this. If you need multithreaded data loading, this should be a different abstraction, but as far as I see everything you mentioned seems possible to implement with `DataLoader`.

thanks for the example!

I am just wondering what's the last line of the code doing? when and why do we need to call `finish_sequence()`? If without it, will the dynamic loader work as intended?

Thanks!
